src/core/net/ip4_address.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for IPv4 addresses.
+ */
+
+#ifndef IP4_ADDRESS_HPP_
+#define IP4_ADDRESS_HPP_
+
+#include "openthread-core-config.h"
+
+#include "common/clearable.hpp"
+#include "common/equatable.hpp"
+#include "common/error.hpp"
+#include "common/string.hpp"
+
+namespace ot {
+namespace Ip4 {
+
+/**
+ * This class represents an IPv4 address.
+ *
+ */
+OT_TOOL_PACKED_BEGIN
+class Address : public Equatable<Address>, public Clearable<Address>
+{
+public:
+    enum : uint8_t
+    {
+        kSize              = 4,  ///< Size of an IPv4 Address (in bytes).
+        kAddressStringSize = 17, ///< Max string size to represent an IPv4 address (`ToString()`).
```

Comment:
Is the worst case 16 chars (including null char) - e.g. 255.255.255.255?

Reply:
Yes. The worst-case string length we would need indeed `16`. 
But I went for `17` here intentionally as a small code-size optimization 😃 
- Every time we use `String<size>` with some size value, a new instance of the template is created (along with its method). Though they are small, they still do increase code size.
- If I see that the info string size for a new type is just few bytes different from an existing size that we already use for some other type, I use that instead to share the `String<size>` template instantiation (and reduce code size).
- We do have some of other types that use 17 as their `kInfoStringSize`.
- I have used this trick in some other cases, we have a bunch of types that use `45` as their string size.

Reply:
Nice! Maybe update the comment to clarify that this is not really the "max string size to represent an IPv4 address"?